bazel: TestExternal tests in pkg/sql/opt/xform fail with --rewrite

When I run `./dev test ./pkg/sql/opt/xform -f 'TestExternal' --rewrite`, I see a handful of errors like:

```
--- FAIL: TestExternal/ycsb (0.00s)
        opt_tester.go:1800: open /private/var/tmp/_bazel_marcus/61eb653382e9435773303080c4f9ae5d/sandbox/darwin-sandbox/181/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin-fastbuild/bin/pkg/sql/opt/xform/xform_test_/xform_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/testutils/opttester/testfixtures/ycsb_schema: operation not permitted
```

This is likey because this tests tries to rewrite files outside of its own testdata directory, which requires special handling.
looks like this test is trying to --rewrite files outside of its own testdata directory. these cases need special handling (for example, see https://github.com/cockroachdb/cockroach/commit/c1f28774675372eb56fd79348ee9181fa7ac9705).

Epic CRDB-8036

--- a/dev
+++ b/dev
@@ -3,7 +3,7 @@
 set -euo pipefail
 
 # Bump this counter to force rebuilding `dev` on all machines.
-DEV_VERSION=31
+DEV_VERSION=32
 
 THIS_DIR=$(cd "$(dirname "$0")" && pwd)
 BINARY_DIR=$THIS_DIR/bin/dev-versions

--- a/pkg/cmd/dev/test.go
+++ b/pkg/cmd/dev/test.go
@@ -105,6 +105,14 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 		showLogs      = mustGetFlagBool(cmd, showLogsFlag)
 		count         = mustGetFlagInt(cmd, countFlag)
 		vModule       = mustGetFlagString(cmd, vModuleFlag)
+
+		// These are tests that require access to another directory for
+		// --rewrite.
+		extraRewritablePaths = []struct{ pkg, path string }{
+			{"pkg/ccl/logictestccl", "pkg/sql/logictest"},
+			{"pkg/sql/opt/memo", "pkg/sql/opt/testutils/opttester/testfixtures"},
+			{"pkg/sql/opt/xform", "pkg/sql/opt/testutils/opttester/testfixtures"},
+		}
 	)
 
 	var disableTestSharding bool
@@ -175,11 +183,13 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 		for _, testTarget := range testTargets {
 			dir := getDirectoryFromTarget(testTarget)
 			args = append(args, fmt.Sprintf("--sandbox_writable_path=%s", filepath.Join(workspace, dir)))
-			if strings.Contains(testTarget, "pkg/ccl/logictestccl") {
-				// The ccl logictest target shares the testdata directory with the base
-				// logictest target -- make an allowance explicitly for that.
-				args = append(args, fmt.Sprintf("--sandbox_writable_path=%s",
-					filepath.Join(workspace, "pkg/sql/logictest")))
+			for _, extraRewritablePath := range extraRewritablePaths {
+				if strings.Contains(testTarget, extraRewritablePath.pkg) {
+					// Some targets need special handling if they rewrite outside of
+					// their own testdata directory.
+					args = append(args, fmt.Sprintf("--sandbox_writable_path=%s",
+						filepath.Join(workspace, extraRewritablePath.path)))
+				}
 			}
 		}
 	}

--- a/pkg/cmd/dev/testdata/datadriven/test
+++ b/pkg/cmd/dev/testdata/datadriven/test
@@ -121,3 +121,9 @@ dev test pkg/sql/schemachanger --rewrite -v
 ----
 bazel info workspace --color=no
 bazel test pkg/sql/schemachanger:all --nocache_test_results --test_env=GOTRACEBACK=all --test_env=COCKROACH_WORKSPACE=crdb-checkout --test_arg -rewrite --sandbox_writable_path=crdb-checkout/pkg/sql/schemachanger --test_arg -test.v --test_sharding_strategy=disabled --test_output all
+
+exec
+dev test pkg/sql/opt/xform --rewrite
+----
+bazel info workspace --color=no
+bazel test pkg/sql/opt/xform:all --nocache_test_results --test_env=GOTRACEBACK=all --test_env=COCKROACH_WORKSPACE=crdb-checkout --test_arg -rewrite --sandbox_writable_path=crdb-checkout/pkg/sql/opt/xform --sandbox_writable_path=crdb-checkout/pkg/sql/opt/testutils/opttester/testfixtures --test_sharding_strategy=disabled --test_output errors

--- a/pkg/sql/opt/testutils/opttester/BUILD.bazel
+++ b/pkg/sql/opt/testutils/opttester/BUILD.bazel
@@ -43,6 +43,7 @@ go_library(
         "//pkg/sql/sem/tree",
         "//pkg/sql/sem/volatility",
         "//pkg/sql/stats",
+        "//pkg/testutils",
         "//pkg/testutils/sqlutils",
         "//pkg/util",
         "//pkg/util/hlc",
@@ -52,6 +53,7 @@ go_library(
         "//pkg/util/treeprinter",
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_cockroachdb_errors//:errors",
+        "@com_github_cockroachdb_errors//oserror",
         "@com_github_pmezard_go_difflib//difflib",
     ],
 )

--- a/pkg/sql/opt/testutils/opttester/opt_tester.go
+++ b/pkg/sql/opt/testutils/opttester/opt_tester.go
@@ -22,6 +22,7 @@ import (
 	"io/ioutil"
 	"math"
 	"net/url"
+	"os"
 	"path/filepath"
 	"runtime"
 	"sort"
@@ -59,6 +60,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/volatility"
 	"github.com/cockroachdb/cockroach/pkg/sql/stats"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -67,6 +69,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/treeprinter"
 	"github.com/cockroachdb/datadriven"
 	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/errors/oserror"
 	"github.com/pmezard/go-difflib/difflib"
 )
 
@@ -1806,8 +1809,8 @@ func (ot *OptTester) Import(tb testing.TB) {
 // testFixturePath returns the path of a fixture inside opttester/testfixtures.
 func (ot *OptTester) testFixturePath(tb testing.TB, file string) string {
 	if bazel.BuiltWithBazel() {
-		runfile, err := bazel.Runfile("pkg/sql/opt/testutils/opttester/testfixtures/" + file)
-		if err != nil {
+		runfile := testutils.RewritableDataPath(tb, "pkg", "sql", "opt", "testutils", "opttester", "testfixtures", file)
+		if _, err := os.Stat(runfile); oserror.IsNotExist(err) {
 			tb.Fatalf("%s; is your package missing a dependency on \"//pkg/sql/opt/testutils/opttester:testfixtures\"?", err)
 		}
 		return runfile

--- a/pkg/testutils/data_path.go
+++ b/pkg/testutils/data_path.go
@@ -13,6 +13,7 @@ package testutils
 import (
 	"path"
 	"path/filepath"
+	"runtime"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/build/bazel"
@@ -30,14 +31,8 @@ func TestDataPath(t testing.TB, relative ...string) string {
 	// dev notifies the library that the test is running in a subdirectory of the
 	// workspace with the environment variable below.
 	if bazel.BuiltWithBazel() {
-		//lint:ignore SA4006 apparently a linter bug.
-		cockroachWorkspace, set := envutil.EnvString("COCKROACH_WORKSPACE", 0)
-		if set {
-			return path.Join(cockroachWorkspace, bazel.RelativeTestTargetPath(), path.Join(relative...))
-		}
-		runfiles, err := bazel.RunfilesPath()
-		require.NoError(t, err)
-		return path.Join(runfiles, bazel.RelativeTestTargetPath(), path.Join(relative...))
+		relative = append([]string{bazel.RelativeTestTargetPath()}, relative...)
+		return RewritableDataPath(t, relative...)
 	}
 
 	// Otherwise we're in the package directory and can just return a relative path.
@@ -45,4 +40,31 @@ func TestDataPath(t testing.TB, relative ...string) string {
 	ret, err := filepath.Abs(ret)
 	require.NoError(t, err)
 	return ret
+}
+
+// RewritableDataPath returns a path to an asset relative to the top of the
+// workspace. Generally you should use TestDataPath if you're trying to access
+// a file in your test's `testdata` directory, or bazel.Runfile if a read-only
+// link to the file is OK. This function is only necessary if you need the path
+// to a file that you can --rewrite.
+func RewritableDataPath(t testing.TB, relative ...string) string {
+	if bazel.BuiltWithBazel() {
+		cockroachWorkspace, set := envutil.EnvString("COCKROACH_WORKSPACE", 0)
+		if set {
+			return path.Join(cockroachWorkspace, path.Join(relative...))
+		}
+		runfiles, err := bazel.RunfilesPath()
+		require.NoError(t, err)
+		relative = append([]string{runfiles}, relative...)
+		return filepath.Join(relative...)
+	}
+
+	// Get the path to this file from the runtime.
+	_, thisFilePath, _, ok := runtime.Caller(0)
+	if !ok {
+		t.Fatal("unable to get caller information")
+	}
+	cockroachWorkspace := filepath.Dir(filepath.Dir(filepath.Dir(thisFilePath)))
+	relative = append([]string{cockroachWorkspace}, relative...)
+	return filepath.Join(relative...)
 }

